### PR TITLE
[DRAFT] Custom gpu ops custom partitioning

### PR DESCRIPTION
**Pull request by nouiz**
This in progress PR modify the docs/Custom_Operation_for_GPUs.py tutorial to use custom_partitioning instead of xmap.

Don't review now, there is still much work done.
- [x] finish the forward code.
- [x] finish the backward code.
- [ ] update the documentation to explain it.
- [ ] document debugging trick, like looking the the dump after sharding propagation, but before partitioning: module_0019.pjit__unnamed_function_.0011.spmd-partitioner.after_sharding-propagation.before_spmd-partitioning.txt, XLA_FLAGS="--xla_dump_to=DIRECTORY_PATH --xla_dump_hlo_pass_re=.*"
- [ ] commit the full code to easy starting from it.
- [ ] make a test of the commited full code to prevent breaking it as it already happened.
- [ ] modify the documentation to inline part of the commited code instead of duplicating it.

### CODE REVIEW

File: docs/Custom_Operation_for_GPUs.md [hunk 1]
```diff
@@ -633,7 +633,73 @@ With this modification, the `all-gather` operation is eliminated and the custom
...
+
+The partition() function will do a few things:
+- tell which input sharding will be expected. XLA will reshad if needed.
+- tell the final version of the output sharding.
+- give a function that will create the new instruction from the sharded inputs.
+
+See the code comments for more explanation:
+
+```python
+    def infer_sharding_from_operands(eps, mesh, arg_infos, result_infos):
```
Comment (keshavb96) [inline docs/Custom_Operation_for_GPUs.md:669]: Some type annotations would be quite useful here, especially for people not entirely familiar with the signatures and expected arguments. For example:

`def infer_sharing_from_operands(
eps : float, mesh : jax.sharding.Mesh, arg_infos : <whatever the type>, result_infos : <whatever the type>
):`

File: docs/Custom_Operation_for_GPUs.md [hunk 2]
```diff
@@ -633,7 +633,73 @@ With this modification, the `all-gather` operation is eliminated and the custom
...
+        del eps, result_infos
+        x_info, weight_info = arg_infos
```
Comment (keshavb96) [inline docs/Custom_Operation_for_GPUs.md:671]: Again type annotations on what `x_info` and `weight_info` would be useful. But perhaps if `arg_infos` was annotated in the function signature we wouldn't need these?

File: docs/Custom_Operation_for_GPUs.md [hunk 3]
```diff
@@ -633,7 +633,73 @@ With this modification, the `all-gather` operation is eliminated and the custom
...
+        assert len(x_info.shape) == 3
+        assert len(weight_info.shape) == 2
+        # partition() will force all dims of all inputs to be replicated except the
+        # first dim of x that will be kept as is.
+	# This is because the implementaion we can only shard on the batch dimensions.
+        x_spec = get_padded_spec(arg_infos[0])
```
Comment (keshavb96) [inline docs/Custom_Operation_for_GPUs.md:677]: A concrete example of what `get_padded_spec` is doing would be very useful

File: docs/Custom_Operation_for_GPUs.md [hunk 4]
```diff
@@ -633,7 +633,73 @@ With this modification, the `all-gather` operation is eliminated and the custom
...
+The partition() function will do a few things:
+- tell which input sharding will be expected. XLA will reshad if needed.
+- tell the final version of the output sharding.
+- give a function that will create the new instruction from the sharded inputs.
+
+See the code comments for more explanation:
+
+```python
+    def infer_sharding_from_operands(eps, mesh, arg_infos, result_infos):
+        del eps, result_infos
```
Comment (keshavb96) [inline docs/Custom_Operation_for_GPUs.md:670]: What is the reason for this deletion? If they are unused in the function, why pass them in? The reason might be something simple for someone who is already familiar with this pattern, but for someone who isn't as familiar this seems strange, especially since this pattern isn't discussed in the documentation on [custom partitioning](https://jax.readthedocs.io/en/latest/jax.experimental.custom_partitioning.html#module-jax.experimental.custom_partitioning)

**Comment by nouiz** [comment]
Replaced by https://github.com/google/jax/pull/20179

**Comment by nouiz** [comment]
Replaced by https://github.com/google/jax/pull/20179
